### ISSUE TITLE
An alternative is LittlevGL

### ISSUE THREAD

**Issue by bkauler** [kakaroto/fbwhiptail#1]
This is just an info post, you can close it at any time.

I have just posted to my blog about fbwhiptail:

http://bkhome.org/news/201809/fbwhiptail-whiptail-for-the-framebuffer.html

As I commented, it might be OK for the needs at Purism, but if you want something more, it is worth considering LittlevGL. You can create all kinds of sophisticated GUIs.

I have only had a fairly brief experience playing with LittlevGL, and I hit one problem: without X (and hence without sdl2), input is provided by evdev and screen output by the framebuffer. However, currently only mouse input works for evdev. I can see that a solution could be worked out, and made a suggestion, but as unfamiliar with the details of how LVGL works, and having other projects to get back to, left it as a proposal for someone to take up:

https://github.com/littlevgl/lvgl/issues/377

LVGL project page:

https://littlevgl.com/

**Comment by kakaroto** [comment]
Hi,

Thanks for the info!
I know there are other more fully developed framebuffer GUI libraries out there, and littlevGL seems interesting. I was also suggested Nuklear (https://github.com/vurtun/nuklear) which also looks great.
fbwhiptail is different for a few reasons, and I read your "review" in your blog post, and yes, it's currently very limited as you've noticed.

The story behind it is two fold : 
First, this was a project that I wrote for fun during a single weekend just to see "if it would work". Most/all of the cairo-based menu library I wrote in [2011](https://github.com/kakaroto/SGTPuzzles/blob/master/ps3/ps3menu.c) just for displaying the menu of a homebrew game I developed for the PS3 which was a port the [SGT Puzzles collection](https://www.chiark.greenend.org.uk/~sgtatham/puzzles/) to the jailbroken PS3. At that time, the only 'graphics' library that was ported and working on the PS3 was Cairo (we didn't even have an openGL implementation written yet, or even thread capabilities, the homebrew toolchain was very young still..). So this project was born from the curiosity of "Can I reuse that library that I wrote in 2011 and quickly make it show something that looks like whiptail?" and so within a few hours, I wrote fbwhiptail and it worked and my curiosity transformed into "it's usable enough so let's use it as a whiptail replacement in Heads".
After that, I never really had any time to improve it into something better and there wasn't any need for me to spend more time on it either. If Purism needs something from me done, like "we need yes/no support", I add it quickly or "we want the theme to be different" I quickly change it, or "our string doesn't fit", I can tweak the currently hardcoded sizes/widths/etc.. then continue working on my main project.
So this is, basically, a proof of concept with the plan for all the missing functionality to be added when time allows or the need arises. Funnily enough though this proof of concept is solid enough to be used in Heads and it already fills all of their needs.

The second part of the story is that, I suggested writing a "real GUI", something in C that is customized for us, etc.. but I was told no, and the reason is simple, most of the people who would be contributing to the project (and I think the main person working on the GUI part) would be much more proficient in Bash than in C, and people would know how to use whiptail in the various scripts that Heads has, rather than have them learn some new library, and write the functionality in C, etc.. The other reasoning is that not every machine that Heads would run on will support having a framebuffer, so it's currently very convenient to create the GUI with whiptail and have it work on every machine, then if your machine supports the framebuffer, you just enable fbwhiptail and your GUI is suddenly graphically more pleasant, without needing to change anything in all of the scripts of Heads.
This is why I was told to keep this as a simple whiptail replacement, as a simple program rather than a complex GUI program, or a library.

> Interesting, fbwhiptail compiles to 'fbwhiptail' and 'gtkwhiptail', so will also work with cairo and gtk

Actually, the 'gtkwhiptail' is just a gtk wrapper so I can test my changes on the PC without having to switch to a tty console, or test on a different machine. It's also useful I guess for people like you to see what fbwhitail looks like easily.

> I have tested it... there is a lot of functionality missing, compared with whiptail.  Most of the dialogs are not yet implemented. I tested a simple "--msgbox" and "--yesno". There does not seem to be any automatic text wrap, the <height> and <width> parameters seem to do nothing, text-input only, my mouse did not work. There seems to be a fixed theme. 

Yeah, we only have/need --msgbox,  --yesno, --menu and --gauge.
There is no automatic text wrap, but you can use \n in your string for the text header, and it will split those into multiple lines. The sizes/heights/widths/etc.. are hardcoded [here](https://github.com/kakaroto/fbwhiptail/blob/master/fbwhiptail_menu.h#L95). Changing the code to take those as arguments and automatically calculate the best values would be my first TODO item when I get back to this project. 
There is no mouse support because it's not really needed, and it's too complex to add for very little gain, but mostly because Heads doesn't have mouse drivers in the kernel as it needs to load the minimum required drivers to have as little verified code running (something like that), so I don't see the point.
As for the theme, I do have some extra options that are unique to fbwhiptail which can be used to 'configure the theme' : 
```
  Frambuffer options:
           --background-png <file>           Display PNG image as background
           --background-gradient <start red> <start green> <start blue> <end red> <end green> <end blue>
                                             Generate a linear gradient background from left to right
          --text-size <size>                 Set the text font size
```
The current "theme" is actually the same one that I used for the PS3 game but with the color turned to grey instead of bright blue
You can see images of the original game here :  ![here](http://kakaroto.homelinux.net/wp-content/uploads/2011/05/IMG_9975-538x358.jpg) and ![here](http://kakaroto.homelinux.net/wp-content/uploads/2011/05/IMG_9976-538x358.jpg).
What I like about the menu API I wrote is that it's very customizable, you can use images for defining the buttons, but also (as is the case in fbwhiptail) the entire menu button is programatically generated. So you could change the color scheme by simply changing these two lines  [here](https://github.com/kakaroto/fbwhiptail/blob/master/fbwhiptail_menu.c#L297) in the create_standard_background function :
```
  cairo_pattern_add_color_stop_rgb (linpat, 0.0, 0.3, 0.3, 0.3);
  cairo_pattern_add_color_stop_rgb (linpat, 1.0, 0.7, 0.7, 0.7);
```

So basically, you could think of it as this being a very simple cairo-based library with a single widget: a menu, and fbwhiptail is a C program that uses that same menu API to create complex uses of the menu widget. The same fbwhiptail could probably be made using either of the other existing framebuffer UI libraries, but fbwhiptail takes only about 200KB more than whiptail itself (including cairo and all its dependencies), I'd assume that a more full fledged UI library would take more space than that, and Heads has very limited space to work with on the SPI flash.

p.s: I should probably summarize this in a README.. at least make it clear it's more of a working proof of concept than a fully developped app...

**Comment by bkauler** [comment]
That is a lovely detailed reply!

I have posted a link from my blog to this issue #1, so that anyone who reads that post on my blog will get the full picture:

http://bkhome.org/news/201809/fbwhiptail-whiptail-for-the-framebuffer.html

**Comment by kakaroto** [comment]
Thanks! Glad you enjoyed reading it and thanks for updating the blog post :)
